scipy/sparse/csgraph/_matching.pyx

```diff
@@ -273,3 +274,495 @@ cdef tuple _hopcroft_karp(ITYPE_t[:] indices, ITYPE_t[:] indptr,
+                if cc[t] < v[jp]:
+                    v[jp] = cc[t]
+                    y[jp] = i
+        for jp in range(nc - 1, -1, -1):
+            i = y[jp]
+            if x[i] == -1:
+                x[i] = jp
+            else:
+                y[jp] = -1
+                # Here, the original Pascal code simply inverts the sign of x;
+                # as that doesn't play too well with zero-indexing, we
+                # explicitly keep track of uniqueness instead.
+                xinv[i] = 1
+        lp = 0
+        for i in range(nr):
+            if xinv[i] == 1:
+                continue
+            if x[i] != -1:
+                min_diff = INFINITY
+                j1 = x[i]
+                for t in range(first[i], first[i + 1]):
+                    jp = kk[t]
+                    if jp != j1:
+                        if cc[t] - v[jp] < min_diff:
+                            min_diff = cc[t] - v[jp]
+                u[i] = min_diff
+                tp = first[i]
+                while kk[tp] != j1:
+                    tp += 1
+                v[j1] = cc[tp] - min_diff
+            else:
+                free[lp] = i
+                lp += 1
+        for _ in range(2):
+            h = 0
+            l0p = lp
+            lp = 0
+            while h < l0p:
+                i = free[h]
+                h += 1
+                # Note: In the original Pascal code, the indices of the lowest
+                # and second-lowest reduced costs are never reset. This can
+                # cause issues for infeasible problems; see
+                # https://stackoverflow.com/q/62875232/5085211
+                j0p = -1
+                j1p = -1
+                v0 = INFINITY
+                vj = INFINITY
+                for t in range(first[i], first[i + 1]):
+                    jp = kk[t]
+                    dj = cc[t] - v[jp]
+                    if dj < vj:
+                        if dj >= v0:
+                            vj = dj
+                            j1p = jp
+                        else:
+                            vj = v0
+                            v0 = dj
+                            j1p = j0p
+                            j0p = jp
+                # If the index of the column with the largest reduced cost has
+                # not been set, no assignment is possible for this row.
+                if j0p < 0:
+                    raise ValueError('no full matching exists')
+                i0 = y[j0p]
+                u[i] = vj
+                if v0 < vj:
+                    v[j0p] += v0 - vj
+                elif i0 != -1:
+                    j0p = j1p
+                    i0 = y[j0p]
+                x[i] = j0p
+                y[j0p] = i
+                if i0 != -1:
+                    if v0 < vj:
+                        h -= 1
+                        free[h] = i0
+                    else:
+                        free[lp] = i0
+                        lp += 1
+        l0 = lp
```

Comment:
Seems to be missing this line (105) from pascal
h:=0; for i:=1 to n do h:=h+u[i]+v[i]. 
however, I this value of h doesnt seem to be used before h is reassigned.

Reply:
Yep, good catch. And indeed, I reached the same conclusion: That loop appears to be dead code.